quinn-proto/src/connection/assembler.rs

```diff
@@ -59,6 +59,39 @@ impl Assembler {
         Some((n, data))
     }
 
+    // Get the the next ordered chunk
+    pub(crate) fn read_chunk(&mut self) -> Result<Option<Bytes>, IllegalOrderedRead> {
+        if let State::Unordered { .. } = self.state {
+            return Err(IllegalOrderedRead);
+        }
+
+        loop {
+            let chunk = match self.data.peek() {
+                Some(chunk) => chunk,
+                None => return Ok(None),
+            };
+
+            if chunk.offset > self.bytes_read {
+                // Next chunk is after current read index
+                return Ok(None);
+            } else if (chunk.offset + chunk.bytes.len() as u64) <= self.bytes_read {
+                // Next chunk is useless as the read index is beyond its end
+                drop(chunk);
+                self.data.pop();
+                continue;
+            }
```

Comment:
Is this actually necessary? I thought we prevented this scenario in the assembler.

Reply:
It replicates the logic I found in `consume()`.

Reply:
Looks like `insert()` doesn't push already received data, but a consecutive insert containing the same data would leave the precedent useless chunk in (unless defragmentation happens ?).

Reply:
This case is needed per https://github.com/quinn-rs/quinn/pull/952#discussion_r549865148.